packages/gatsby-plugin-sharp/src/worker.js

```diff
@@ -0,0 +1,48 @@
+const queue = require(`async/queue`)
+const { processFile } = require(`./process-file`)
+
+/** @typedef {import('./process-file').TransformArgs} TransformArgs */
+
+/**
+ * @typedef WorkerInput
+ * @property {string} inputPath the file path to transform
+ * @property {string} contentDigest
+ * @property {object} pluginOptions
+ * @property {{outputPath: string, transforms: TransformArgs[]}} operations
+ */
+
+/**
+ * the queue concurrency is 1 as we only want to transform 1 file at a time.
+ * @param {(job: WorkerInput, callback: Function) => undefined} task
+ */
+const q = queue((job, callback) => {
+  Promise.all(
+    processFile(
+      job.inputPath,
+      job.contentDigest,
+      job.operations.map(operation => {
+        return {
+          outputPath: operation.outputPath,
+          args: operation.transforms,
+        }
+      }),
+      job.pluginOptions
+    )
+  )
+    .then(() => callback())
+    .catch(err => callback(err))
+}, 1)
+
+/**
+ * @param {WorkerInput} job
+ */
+module.exports = job =>
```

Comment:
Why does the worker have a queue? My understanding of what we're building is that workers are ephemeral (cloud functions can stop working at any time) so they can die on errors, etc. so state shouldn't be stored in the worker.

Reply:
Sadly we can't move the queue outside of a worker as it needs to know how many jobs it can execute at once. If we want to move it outside of the worker, we have to add concurrency to our jobs. This is hard because concurrency might vary on cloud vs local. We could say concurrency for a job is only used locally, and Infinity is used when doing cloud. Than it's up to cloud's pubsub to handle it.
```suggestion
module.exports = job =>
```
```suggestion
module.exports = job =>
```
```suggestion
module.exports = job =>
```

I'll give an example with images. Sharp is currently set to only process 1 inputFile and as many transforms of that file as it can for performance reasons. On cloud, we technically have an infinity amount of concurrency. When we put the concurrency in the worker we fix this problem as the worker knows how many it can execute at once. For local this means we only run one file at the time. For cloud, one function can run one input file at the time, which still gives us an infinity amount of parallel execution.

If we want to move it outside of the worker we'll probably want to change the api to something like
```
const queue = createJobQueue('IMAGE_PROCESSING', { concurrency: 1 })
queue .enqueue(args)
```
As in my opinion `enqueueJob('IMAGE_PROCESSING', args, { concurrency: 1})` is weird as each job potentially could change concurrency.

Does this make sense?

Reply:
Sadly we can't move the queue outside of a worker as it needs to know how many jobs it can execute at once. If we want to move it outside of the worker, we have to add concurrency to our jobs. This is hard because concurrency might vary on cloud vs local. We could say concurrency for a job is only used locally, and Infinity is used when doing cloud. Than it's up to cloud's pubsub to handle it.

I'll give an example with images. Sharp is currently set to only process 1 inputFile and as many transforms of that file as it can for performance reasons. On cloud, we technically have an infinity amount of concurrency. When we put the concurrency in the worker we fix this problem as the worker knows how many it can execute at once. For local this means we only run one file at the time. For cloud, one function can run one input file at the time, which still gives us an infinity amount of parallel execution.

If we want to move it outside of the worker we'll probably want to change the api to something like
```
const queue = createJobQueue('IMAGE_PROCESSING', { concurrency: 1 })
queue.enqueue(args)
```
As in my opinion `enqueueJob('IMAGE_PROCESSING', args, { concurrency: 1})` is weird as each job potentially could change concurrency.

Does this make sense?

Reply:
Resolving this for now, we'll revisit this later if necessary